nixos/beszel: init

BonusPlay opened the pull request:
Beszel is a lightweight server monitoring platform that includes Docker statistics, historical data, and alert functions. It normally is deployed using docker, but due to golang, it's quite easy to deploy as systemd service as well.

While I've been testing this module in my homelab, I've noticed #379932 popped up, but that creates only agent (and is marked as WIP), while this has both agent + hub.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

arunoruto commented on nixos/modules/services/monitoring/beszel-agent.nix:
```diff
@@ -0,0 +1,107 @@
+{
+  meta.maintainers = with lib.maintainers; [ BonusPlay ];
+
+  options.services.beszel-agent = {
```
More of a question than a review: Would it be better to have options under services.beszel.agent instead of services.beszel-agent, and the same for the hub?

BonusPlay replied:
I've pushed commits with this change, however they are not tested yet. I'll deploy them to my infra for testing soon-ish (tomorrow?)

arunoruto replied:
Amazing! I guess in the future a service option `services.beszel.enable` could enable the agent and hub in one go, if the default options allow it somehow.
But that can be left for a different PR :)

SuperSandro2000 commented on nixos/modules/services/monitoring/beszel-agent.nix:
```diff
@@ -0,0 +1,107 @@
+      '';
+    };
+    extraPath = lib.mkOption {
+      type = lib.types.nullOr (lib.types.listOf lib.types.package);
+      default = null;
```
```suggestion
      type = lib.types.listOf lib.types.package;
      default = [ ];
```

SuperSandro2000 commented on nixos/modules/services/monitoring/beszel-agent.nix:
```diff
@@ -0,0 +1,107 @@
+      after = [ "network-online.target" ];
+
+      environment = cfg.environment;
+      path = lib.optionals (cfg.extraPath != null) cfg.extraPath;
```
```suggestion
      path = cfg.extraPath;
```

SuperSandro2000 commented on nixos/modules/services/monitoring/beszel-agent.nix:
```diff
@@ -0,0 +1,107 @@
+
+      serviceConfig = {
+        ExecStart = ''
+          ${cfg.package}/bin/beszel-agent
```
```suggestion
          ${lib.getExe cfg.package}
```

arunoruto replied:
There is one package `beszel` which unifies the agent and hub binaries. I would also be a fan of packaging the binaries separately. AFAIK `getExe` shouldn't work here and for the hub service 🤔

SuperSandro2000 commented on nixos/modules/services/monitoring/beszel-hub.nix:
```diff
@@ -0,0 +1,126 @@
+
+      serviceConfig = {
+        ExecStart = ''
+          ${cfg.package}/bin/beszel-hub serve --http='${cfg.host}:${toString cfg.port}'
```
```suggestion
          ${lib.getExe cfg.package} serve --http='${cfg.host}:${toString cfg.port}'
```

arunoruto commented on nixos/modules/services/monitoring/beszel-agent.nix:
```diff
@@ -0,0 +1,107 @@
+
+      serviceConfig = {
+        ExecStart = ''
+          ${lib.getExe cfg.package}/bin/beszel-agent
```
```suggestion
          ${cfg.package}/bin/beszel-agent
```
This won't work, since the the `beszel` package has no `mainProgram` attribute. The same goes for the `hub` config. If the beszel ever gets split into an agent and hub package, then you could use `lib.getExe`.

BonusPlay replied:
Not to mention I left out `/bin/beszel-agent` at the end. Btw, should I add you as a maintainer to those modules?

arunoruto replied:
Sure, I do care enough about the project to help out maintaining it :)

NyCodeGHG commented on nixos/modules/services/monitoring/beszel-agent.nix:
```diff
@@ -0,0 +1,107 @@
+
+    package = lib.mkPackageOption pkgs "beszel" { };
+
+    user = lib.mkOption {
```
Is there any use case to running beszel under a different user?
Otherwise I would suggest to remove that option.

BonusPlay replied:
no, but it's useful for setting secret ownership (like environment variables)

NyCodeGHG replied:
You don't need an extra user for that, only root needs to be able to read a file for systemd's EnvironmentFile.
DynamicUser=true doesn't do anything if there is already a user with that name. 
We should use DynamicUser over an extra user if there is no good reason not to.

BonusPlay replied:
I think you're right in this case. It helps when something is config file (loaded during runtime) and not env file. Applied suggestion :+1:.

NyCodeGHG commented on nixos/modules/services/monitoring/beszel-agent.nix:
```diff
@@ -0,0 +1,107 @@
+    };
+
+    users.groups.${cfg.group} = { };
+    users.users.${cfg.user} = {
```
We don't need to create the user ourselves, if we use DynamicUser=true

arunoruto commented on nixos/modules/services/monitoring/beszel-agent.nix:
```diff
@@ -0,0 +1,89 @@
+        DynamicUser = true;
+        LockPersonality = true;
+        NoNewPrivileges = true;
+        PrivateDevices = true;
```
This causes some problems with running `nvidia-smi` with the service. I always get, that nvidia-smi can't find GPUs on the system. I guess it can be set to false? Or set it to false, if the GPU environmental variable is turned on!

BonusPlay replied:
Hmm, it could be a trick if some users decide to use only `environmentFile` (which I don't think we can handle). I'm also hesitant to create dedicated `gpu` switch (but maybe that is the way?). WDYT

NyCodeGHG replied:
I think it would make sense to turn PrivateDevices off, as it clearly seems to breaks a feature

arunoruto replied:
> Hmm, it could be a trick if some users decide to use only `environmentFile` (which I don't think we can handle). I'm also hesitant to create dedicated `gpu` switch (but maybe that is the way?). WDYT

The options are:
1. Remove the option if it isn't critical. Maybe you can explain why you added it since it wasn't present in the default systemd config provided by beszel.
2. Remove the env file option and only provide the env option. Look if the GPU attribute is provided, and if it is, check if it is `true` or `"true"`. This is much more complicated, but if the option is critical, we could keep it :)

BonusPlay replied:
Just copy&paste of other modules as templates. I will remove it.

arunoruto commented on nixos/modules/services/monitoring/beszel-agent.nix:
```diff
@@ -0,0 +1,89 @@
+      type = lib.types.listOf lib.types.package;
+      default = [ ];
+      description = ''
+        Extra packages to add to beszel path (such as nvidia-smi or rocm-smi).
```
Maybe give an example how to add nvidia-smi and rocm-smi. Before starting to use this service implementation, I added `/run/current-system/sw` to extra-path (which worked). No I had to provide a package/derivation?, so I had to add the following item to the list
`(lib.getBin config.boot.kernelPackages.nvidiaPackages.stable)`

BonusPlay replied:
I think we could check for nvidia/amd GPU packages being enabled and pre-emptively  add them to the path.

arunoruto replied:
Thats one way, check if `amdgpu` or `nvidia` are inside `config.services.xserver.videoDrivers`, and add the appropriate package.
The other option is to add the `/run/current-system/sw` path to `path`:
```nix
path = [ "/run/current-system/sw" ] ++ cfg.extraPath;
```
But I am not sure if this is desired, even tho it worked in my previous setup...

arunoruto commented on nixos/modules/services/monitoring/beszel-agent.nix:
```diff
@@ -0,0 +1,95 @@
+      environment = cfg.environment;
+      path =
+        cfg.extraPath
+        ++ (builtins.elem "nvidia" config.services.xserver.videoDrivers) [
```
```suggestion
        ++ lib.optionals (builtins.elem "nvidia" config.services.xserver.videoDrivers) [
```
You forgot the `lib.optionals` here.

arunoruto commented on nixos/modules/services/monitoring/beszel-agent.nix:
```diff
@@ -0,0 +1,95 @@
+    arunoruto
+  ];
+
+  options.services.beszel.agent = {
```
I just ran the agent on a system with an AMD iGPU to test the `rocm-smi` integration. It seems to work, but only when the correct port has been opened. I suggest to add an `openFirewall` option:

```nix
# in the options sections
openFirewall = lib.mkEnableOption "Open the firewall port 45876.";
# and add the following in the config sections
networking.firewall.allowedTCPPorts = lib.mkIf cfg.openFirewall [
  45876
];
```

The only complicated thing is, users could specify a custom port via the environment setting and the variable `PORT`. Checking if the key `PORT` is present in `cfg.environment` and putting it in the list, instead of `45876` would be the best option, and I will look into it later :)

BonusPlay replied:
I haven't tested the commit, but iirc `cfg.environment` has to be a string, so I added `toInt` :)

arunoruto reviewed (approved):
I have been using the config for a while now and haven't had any problems so far. IMO it can be merged :)

jolheiser reviewed (approved):
I've put this on a handful of machines in my setup as well with no issues fwiw :+1:

arunoruto commented on nixos/modules/services/monitoring/beszel-hub.nix:
```diff
@@ -0,0 +1,108 @@
+        PrivateUsers = true;
+        ProtectClock = true;
+        ProtectControlGroups = true;
+        ProtectHome = true;
```
I just checked what the option means (some [reddit knowledge](https://www.reddit.com/r/linuxquestions/comments/chmmg5/systemd_protecthome_what_is_its_benefit/)), and I am not sure why we would give beszel access to the home folder(s)? Using `read-only` could be solving the symptom, but not the underlying problem in this case with your specific config or there is a more appropriate way to set this. Maybe try to set it to `tmpfs`, which emulates the folders, without actually providing access to them (if I am not mistaken).

BonusPlay commented on nixos/modules/services/monitoring/beszel-agent.nix:
```diff
@@ -0,0 +1,104 @@
+
+      environment = cfg.environment;
+      path =
+        cfg.extraPath
```
> `or (builtins.elem "modesetting" config.services.xserver.videoDrivers)`

wouldn't this match nvidia as well?

arunoruto replied:
> wouldn't this match nvidia as well?

I haven't seen it in an Nvidia setting, but I am also not fully sure... I guess just having the intel check would be fine too.

BonusPlay replied:
I've left the intel case. Thanks for the suggestion :+1:

BonusPlay committed:
nixos/beszel.hub: init

Co-authored-by: Mirza Arnaut <mirza.arnaut45@gmail.com>

BonusPlay committed:
nixos/beszel.agent: init

Co-authored-by: Mirza Arnaut <mirza.arnaut45@gmail.com>

BonusPlay commented:
rebased to master, moved released notes to 25.11, still hoping for a merge

arunoruto reviewed (approved):
I have been using the module since day one and it was running without a hitch!
Would be awesome to merge it and fix anything else in separate PRs.

arunoruto commented:
> I have been using this PR for a couple of days, and it works almost well. I have one issue: the agent doesn't report the Docker-related stuff because it doesn't have the right to do so.
> 
> ```
> DEBUG Containers err="Get \"http://localhost/containers/json\": dial unix /var/run/docker.sock: connect: permission denied"
> ```
> 
> I bypassed this by setting it to root, but this is obviously not a good method.

That makes sense, since a dynamic user is being used to launch the service. I found a [reddit post](https://www.reddit.com/r/docker/comments/z5hm75/docker_and_systemd_dynamicuser/) with the same problem, and it seems like using `SupplementaryGroups=docker` solves the problem. If the docker group can be set somewhere with an option, it would be even better to insert it into the option. I am not well versed in docker under NixOS, so maybe someone else can give a better solution. For example, if `[virtualisation.podman.dockerSocket.enable](https://search.nixos.org/options?channel=unstable&query=virtualisation.podman)` is enabled, I am not sure under which group the socket is made available?
But `SupplementaryGroups=docker` should fix the problem and if any further problems arise in the future, a separate PR can be made!

lunik1 commented:
The documentation of the [`virtualisation.podman.dockerSocket.enable`](https://search.nixos.org/options?channel=unstable&query=virtualisation.podman) option states

> Users must be in the `podman` group in order to connect. As with Docker, members of this group can gain root access.

hatch01 reviewed (approved):
Works in production for a week.

h7x4 committed:
nixos/tests/beszel: init

h7x4 commented:
> I know we can polish this module to perfection and bikeshed for another half a year, but this PR has been in decently-working state for over 8 months. I know it could be better, but I'd like to try get this module merged into 25.11 (we've already missed 25.05). @h7x4 is that possible?

Yes, I'd be happy to get it merged by 25.11. It seems like most technical and UX issues with potential for backwards incompatibilities has been ironed out.

It took a little while, but I finished a nixos test. Do you mind if I push it on top of the current patchset?

BonusPlay commented:
I added @arunoruto as co-author to the commits since he proposed a lot of changes. I've also rebased on master and fixed a merge conflict. @h7x4 please check if beszel package has correct `passthru` section.
